c_src/membrane_audio_mix_plugin/mixer.c

```diff
@@ -0,0 +1,242 @@
+void scale(int64_t *values, uint32_t values_length, double quotient,
+           UnifexState *state) {
+
+  for (uint32_t i = 0; i < state->queue_length; ++i) {
+    state->queue[i] = (int64_t)(state->queue[i] * quotient);
+  }
+
+  for (uint32_t i = 0; i < values_length; ++i) {
+    values[i] = (int64_t)(values[i] * quotient);
+  }
+}
+
+/**
+ * Takes given values and values in queue and coverts them to samples.
+ *
+ * If any of the values overflows limits of the format, values will be
+ * scaled down so the peak of the wave will become
+ * maximal (minimal) allowed value.
+ */
+void get_samples(uint8_t *samples, int64_t *values, uint32_t values_length,
+                 UnifexState *state) {
+  int64_t min = state->sample_max;
+  int64_t max = state->sample_min;
+
+  for (uint32_t i = 0; i < state->queue_length; ++i) {
+    if (state->queue[i] < min) {
+      min = state->queue[i];
+    }
+
+    if (state->queue[i] > max) {
+      max = state->queue[i];
+    }
+  }
+
+  for (uint32_t i = 0; i < values_length; ++i) {
+    if (values[i] < min) {
+      min = values[i];
+    }
+
+    if (values[i] > max) {
+      max = values[i];
+    }
+  }
+
+  if (min < state->sample_min) {
+    double quotient = (state->sample_min * 1.0) / min;
+    scale(values, values_length, quotient, state);
+  } else if (max > state->sample_max) {
+    double quotient = (state->sample_max * 1.0) / max;
+    scale(values, values_length, quotient, state);
+  }
+
+  uint8_t *current_sample = samples;
+  for (uint32_t i = 0; i < state->queue_length;
+       ++i, current_sample += state->sample_size) {
+    caps_audio_raw_value_to_sample(state->queue[i], current_sample,
+                                   &state->caps);
+  }
+
+  for (uint32_t i = 0; i < values_length;
+       ++i, current_sample += state->sample_size) {
+    caps_audio_raw_value_to_sample(values[i], current_sample, &state->caps);
+  }
+
+  unifex_free(state->queue);
+  state->queue = NULL;
+  state->queue_length = 0;
+}
+
+/**
+ * Takes given values, divides them into parts where every part must have only
+ * nonnegative or nonpositive values. The whole part consists of values from the
+ * sign change to the sign change. Converts each of these parts into samples -
+ * values might be scaled down to the limit of the format during conversion.
+ *
+ * Any remaining values are put into the state's queue.
+ */
+void chunk_and_scale_to_samples(int64_t *values, uint32_t values_length,
+                                uint8_t *samples, uint32_t *samples_length,
+                                UnifexState *state) {
+  if (values_length == 0 || !values) {
+    return;
+  }
+
+  bool is_wave_positive = state->is_wave_positive;
+  uint32_t start = 0;
+  uint32_t end = next_sign_change(values, values_length, is_wave_positive);
+
+  uint32_t samples_size = 0;
+  while (end < values_length) {
+    uint32_t queue_length = state->queue_length;
+    get_samples(samples + samples_size, values + start, end - start, state);
```

Comment:
By propagating changes I meant renaming that function as well, as it is responsible for `scale_to_samples` part